🏗️(lti_toolbox) accept deep linking LTI request

lebaudantoine opened the pull request:
### Purpose

Enriched the LTI Toolbox capabilities, enabling Content-Item Selection requests, used in deep linking context.

Refined parameter handling, ensuring the LTI class adapts smoothly to varying request types, namely `Launch` or `Selection` ones.

Without these updates, LTI base views could not be used to handle deep linking. `Launch` and `Selection` requests don’t share the same required parameters. Please refer to the LTI documentation for this part.

Plus, some `Launch` parameters should not be passed while sending a `Selection` one.

### Proposition

- [X] Redesign parameter class
- [X] Adapt relevant documentation
- [x] Test the package locally
- [x] Update tests

lebaudantoine committed:
🏗️(lti_toolbox) accept deep linking LTI request

Enriched the LTI Toolbox capabilities, enabling Content-Item Selection
requests, used in deep linking context.

Refined parameter handling, ensuring the LTI class adapts smoothly to
varying request types, namely `Launch` or `Selection` ones.

Without these updates, LTI base views could not be used to handle
deep linking. `Launch` and `Selection` requests don’t share the same
required parameters. Please refer to the LTI documentation for this part.

Plus, some `Launch` parameters should not be passed while sending a
`Selection` one.

lebaudantoine committed:
✏️(lti_toolbox) fix several typos

Found typos, and fixed them.

lebaudantoine committed:
🩹(lti_toolbox) fix double affection of the LTI consumer secret

Found an unusual double variable assignment. After testing the code without
it, I observed flawless functionality, and subsequently fixed the issue.

jmaupetit commented on sandbox/forms.py:
```diff
@@ -7,7 +7,7 @@
 
 class PassportChoiceField(forms.ModelChoiceField):
-    """Select a LTI password"""
+    """Select an LTI password"""
```
sure?

lebaudantoine replied:
Hum, yeah, I would say `a red apple` and not `an red apple`.

jmaupetit commented on src/lti_toolbox/launch_params.py:
```diff
@@ -4,16 +4,17 @@
+LAUNCH_PARAMS_REQUIRED = {"lti_message_type", "lti_version", "resource_link_id"}
 
-LAUNCH_PARAMS_RECOMMENDED = [
+LAUNCH_PARAMS_RECOMMENDED = {
```
Smart :1st_place_medal:

jmaupetit commented on src/lti_toolbox/lti.py:
```diff
@@ -33,37 +29,47 @@ def __init__(self, request):
+    def _process_params(self) -> Union[SelectionParams, LaunchParams]:
+        """Process LTI parameters based on request type."""
+
+        try:
+            is_selection_request = (
+                self.request.POST.get("lti_message_type")
+                == SELECTION_PARAMS_MESSAGE_TYPE
+            )
+
+            if is_selection_request:
+                return SelectionParams(self.request.POST)
+
+            return LaunchParams(self.request.POST)
+
+        except ParamException as error:
+            raise LTIException(f"Exception while processing parameters: {error}")
```
You should reduce to its minimum the scope of try/except blocks to make crystal clear what you are trying to catch:

```suggestion
        is_selection_request = (
            self.request.POST.get("lti_message_type")
            == SELECTION_PARAMS_MESSAGE_TYPE
        )

        try:
            params = SelectionParams(self.request.POST) if is_selection_request else LaunchParams(self.request.POST)
        except ParamException as error:
            raise LTIException(f"Exception while processing parameters: {error}")

        return params
```

lebaudantoine replied:
Improved !

jmaupetit commented on src/lti_toolbox/models.py:
```diff
@@ -129,5 +129,5 @@ def generate_consumer_key() -> str:
     def generate_shared_secret() -> str:
         """Generate a random consumer key."""
         secret_chars = string.ascii_letters + string.digits + "!#$%&*+-=?@^_"
-        secret_size = secret_size = secrets.choice(range(40, 60))
```
Well spotted!

<img src="https://media2.giphy.com/media/iDIJezAxNyRsZTx678/giphy.gif"/>

lebaudantoine replied:
yeah aha

jmaupetit reviewed (commented):
Looks great! :clap:

jmaupetit commented on src/lti_toolbox/lti.py:
```diff
@@ -33,37 +34,50 @@ def __init__(self, request):
+        """Process LTI parameters based on request type."""
+
+        try:
+            is_selection_request = (
+                self.request.POST.get("lti_message_type")
+                == SELECTION_PARAMS_MESSAGE_TYPE
+            )
```
This can be defined before the try/except block right?

lebaudantoine committed:
♻️(lti_toolbox) route captured URL value to `_do_on_success`

While working with the URL dispatcher, it was necessary to capture a value
from the URL using angle brackets. This  captured value was channeled
as an argument into the `post` method.

This method serves as a bridge to the corresponding abstract method,
and I required a seamless way to transmit this captured value from the
`post method` to the final step, denoted as `_do_on_success`.

lebaudantoine committed:
✨(lti_toolbox) add LTIMessageType enum

Introduced a new enumeration encapsulating all LTI message types,
replacing hardcoded constants for improved code organization
and maintainability.

lebaudantoine commented on src/lti_toolbox/launch_params.py:
```diff
@@ -4,16 +4,26 @@
 This is based on lti library (https://github.com/pylti/lti).
```
Should I rename this file to a more generic name, as `params.py`? I didn't as it would be a breaking change.

lebaudantoine commented on src/lti_toolbox/launch_params.py:
```diff
@@ -90,42 +100,63 @@
+    params_required: Set[str] = set()
+    params_is_list: Set[str] = set()
+
     def __init__(self, *args, **kwargs):
```
I've considered adding an extra check, on the LTI message type. Making sure `LaunchParams` has a `"basic-lti-launch-request"` value and `SelectionParams` a `"ContentItemSelectionRequest"` value.

btw, it might be redundant with [this try-except](https://github.com/openfun/django-lti-toolbox/pull/8/files#diff-c1656ed5a48655ee93a36389481047ad3ae291ae94d4e75079874340cf0885d5R34-R45).

Holding off for now due to potential impact on other projects using the library, WDYT?

jmaupetit replied:
Yes!

jmaupetit commented on tests/lti_toolbox/test_launch_params.py:
```diff
@@ -129,3 +126,152 @@ def _launch_params(lti_parameters):
+                }
+            )
+
+    @staticmethod
+    def _selection_params(lti_parameters):
+        consumer = LTIConsumerFactory(slug="test_launch_params")
+        passport = LTIPassportFactory(title="test passport", consumer=consumer)
+        url = "http://testserver/lti/launch"
+        signed_parameters = sign_parameters(passport, lti_parameters, url)
+        return SelectionParams(signed_parameters)
```
For code readability, I would recommend defining this on top of your class (at least before any test case).

lebaudantoine replied:
Totally agree, I kept it as the previous implementation. Lemme refactor both classes.

jmaupetit reviewed (approved):
I only have minor comments.

It's quite surprising to read tests that don't raise anything. I know it was designed this way by previous authors, but still. Surprising :sweat_smile:
